File: 014 Longest Common Prefix.cpp
/*
Longest Common Prefix
https://leetcode.com/problems/longest-common-prefix
Worst time complexity: O(n), n is sum of all string length
Memory Usage: 8.6 MB, less than 100.00% of C++ online submissions
*/
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs)
    {
        if (strs.size() < 1) return "";
        if (strs.size() < 2) return strs[0];
        
        const auto p = strs[0]; // result prefix
        size_t l = p.size(); // result prefix length

        for (const auto& s : strs)
        {
            l = std::min(l, s.size());
            
            for (size_t j = 0; j < l; ++j)
            {
                if (p[j] != s[j])
                {
                    l = j;
                    break;
                }
            }
        }
        return p.substr(0, l);
    }
};

File: 014 Longest Common Prefix.py
'''
14. Longest Common Prefix
https://leetcode.com/problems/longest-common-prefix

Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.

Constraints:
    1 <= strs.length <= 200
    0 <= strs[i].length <= 200
    strs[i] consists of only lowercase English letters.
'''
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ''
    
        common = strs[0]
        commonLength = len(common)

        for word in strs[1:]:
            i = 0;
            while i < min(commonLength, len(word)):
                if (common[i] != word[i]):
                    break;
                i += 1
            commonLength = min(commonLength, i);
          
        return common[:commonLength];
